lib/Sema/SemaDeclAttr.cpp

```diff
@@ -4987,6 +4987,94 @@ static void handleInterruptAttr(Sema &S, Decl *D, const AttributeList &Attr) {
   }
 }
 
+class AMDGPUISAVersionChecker {
+public:
+  AMDGPUISAVersionChecker(Sema &S):S(S){
+    GPU = S.getASTContext().getTargetInfo().getTargetOpts().CPU;
+  }
+
+  /// Check GPU ISA version parameter of AMDGPU attributes.
+  /// The ISA version parameter is either an ISA version string with "_"
+  /// as deliminator, e.g. "8_1_0", or a case insensitive device name,
+  /// e.g. "Fiji".
+  /// \param Attr is AMDGPU attribute containing ISA version parameter.
+  /// \param Index is the index for the ISA version parameter.
+  /// \param ISA will contain the ISA version string if returning true.
+  /// \return true if the attribute does not contain the ISA version parameter,
+  ///         or the ISA version parameter is empty, or the ISA version
+  ///         parameter and the target GPU have the same ISA version.
+  bool checkAMDGPUISAVersion(const AttributeList &Attr, unsigned Index,
+      StringRef &ISA) {
+    if (Attr.getNumArgs() <= Index) {
+      ISA = "";
+      return true;
+    }
+
+    if (!S.checkStringLiteralArgumentAttr(Attr, Index, ISA))
+      return false;
+
+    if (ISA.empty())
+      return true;
+    auto ISAVer = parseAMDGPUISAVersion(ISA);
+    if (ISAVer == AMDGPU_ISA_NONE) {
+      S.Diag(Attr.getLoc(), diag::err_attribute_amdgpu_invalid_isa_version)
+        << ISA;
+      return false;
+    }
+    return ISAVer == parseAMDGPUISAVersion(GPU);
+  }
+
+private:
+  Sema &S;
+  StringRef GPU; // target GPU specified by -target-cpu option
+
+  /// \brief The ISA version of AMD GPU.
+  enum AMDGPUISAVersion {
+    AMDGPU_ISA_NONE,
+    AMDGPU_ISA_600,
+    AMDGPU_ISA_700,
+    AMDGPU_ISA_701,
+    AMDGPU_ISA_800,
+    AMDGPU_ISA_801,
+    AMDGPU_ISA_802,
+    AMDGPU_ISA_803,
+  };
+
+  // Parse AMDGPU ISA version string.
+  // \param ISA is either an ISA version string with "_" as deliminator,
+  //        e.g. "8_1_0", or a case insensitive device name, e.g. "Fiji".
+  // \return AMDGPU ISA version.
+  AMDGPUISAVersion parseAMDGPUISAVersion(StringRef ISA) {
+    return llvm::StringSwitch<AMDGPUISAVersion>(ISA.lower())
```

Comment:
We are moving towards using the GPU arch short names (gfx801, gfx803...) and are phasing out the usage of the code names (fiji, hawaii), could we add the short names as well?